python/pyiceberg/expressions/base.py

```diff
@@ -327,30 +328,26 @@ def bind(self, schema: Schema, case_sensitive: bool) -> BoundIn[T]:
         return BoundIn(bound_ref, tuple(lit.to(bound_ref.field.field_type) for lit in self.literals))  # type: ignore
 
 
-class BooleanExpressionVisitor(Generic[T], ABC):
```

Comment:
I don't think this is a good idea. Don't we want to ensure that all of the cases are handled? What is the return value of visit methods that are not implemented? `None`?

Reply:
Yes it would return `None`. What prompted me to do this is that `BoundBooleanExpressionVisitor` needs to inherit from `BooleanExpression` in order to use the `visit` function defined here. But that means it would need to override each method it doesn't use and return `None` (this is my understanding of what the Java one does).

But I guess there's no reason that needs to be a default behavior from the parent class and it's better to explicitly define that for this visitor. I'll change this back and implement those methods.

Reply:
I would also prefer to keep it abstract. What's also common is to raise a `NotImplementedError`, see https://docs.python.org/3/library/exceptions.html#NotImplementedError. This way you still need to define all the methods, but you can just raise the exceptions on the methods that don't make any sense.

Reply:
Ok updated to revert this back to an ABC class. Since `BoundBooleanExpressionVisitor` is also an ABC class, the abstractmethods are just carried forward. I'll keep in mind the `NotImplementedError` option as we add concrete implementations. 👍

Reply:
> BoundBooleanExpressionVisitor needs to inherit from BooleanExpression in order to use the visit function defined here

I'm not sure I'm following. Did you mean `BooleanExpression`? I don't see why the visitor would need to be an expression.

Reply:
@rdblue sorry that was a mistype, I meant it inherits from `BooleanExpressionVisitor`.